# Testing Models

## (2) KNN

In [ ]:
# import
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Load the dataset - new data set after feature extraction
train = pd.read_csv("new_feauturesExtracted_dataset.csv")

train.head(10)


Id  HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0  1561604                     1                       4    2.0  1388793600   
1   516128                     5                       7    5.0  1033171200   
2  1265775                     0                       0    4.0  1390521600   
3   484323                     3                       4    4.0  1136332800   
4   158350                     1                       2    5.0  1372204800   
5   192763                     5                       8    5.0  1138579200   
6  1060454                     0                       0    4.0  1403654400   
7   860129                     0                       6    3.0  1162080000   
8  1470316                     2                       5    1.0  1357776000   
9   725990                     0                       0    5.0  1396828800   

     neg    neu    pos  compound  Helpfulness  UnHelpfulness  UserAvgScore  \
0  0.153  0.692  0.155   -0.4040     0.250000       0.250000      2.000000   
1  0.000  0.727  0.273    0.9231     0.714286       0.714286      3.083333   
2  0.000  0.767  0.233    0.7901     0.000000       0.000000      4.000000   
3  0.030  0.771  0.199    0.9829     0.750000       0.750000      4.500000   
4  0.000  0.880  0.120    0.4588     0.500000       0.500000      4.600000   
5  0.035  0.833  0.132    0.9938     0.625000       0.625000      4.500000   
6  0.111  0.737  0.152    0.4118     0.000000       0.000000      3.500000   
7  0.151  0.754  0.095   -0.6306     0.000000       0.000000      4.000000   
8  0.030  0.813  0.157    0.7884     0.400000       0.400000      1.000000   
9  0.022  0.800  0.178    0.9504     0.000000       0.000000      5.000000   

   NumUppercase_T  NumUppercase_S  ReviewLength  SummaryLength  Month  Year  \
0               9               2            97              3      1  2014   
1               6               5            43              5      9  2002   
2               0               0            25              1      1  2014   
3              14               1           126              3      1  2006   
4               0               1            23              2      6  2013   
5             188              27           440             17      1  2006   
6               9               2            54              4      6  2014   
7              13               3            46              3     10  2006   
8               3               1            41              1      1  2013   
9              22               1            87              5      4  2014   

   numExclamation_S  numExclamation_T  
0                 1                 0  
1                 1                 0  
2                 0                 1  
3                 1                 3  
4                 0                 0  
5                 0                 4  
6                 0                 0  
7                 0                 0  
8                 0                 0  
9                 0                 0

In [ ]:
# we have to split this data into train and test
# We create a new column called `train` which is `True` if the instance should be included in the training by using the numpy random number generator.

train['train'] = np.random.rand(len(train)) < 0.8

df_train = train[train.train == True]
df_test = train[train.train == False]

print(df_train.shape, df_test.shape)

(111748, 21) (27936, 21)


In [ ]:
train.columns

Index(['Id', 'HelpfulnessNumerator', 'HelpfulnessDenominator', 'Score', 'Time',
       'neg', 'neu', 'pos', 'compound', 'Helpfulness', 'UnHelpfulness',
       'UserAvgScore', 'NumUppercase_T', 'NumUppercase_S', 'ReviewLength',
       'SummaryLength', 'Month', 'Year', 'numExclamation_S',
       'numExclamation_T', 'train'],
      dtype='object')

In [ ]:
predictors = [ 'neg', 'pos', 'compound', 'Helpfulness', 'UnHelpfulness',
       'UserAvgScore', 'NumUppercase_T', 'NumUppercase_S', 'ReviewLength',
       'SummaryLength', 'Month', 'numExclamation_S',
       'numExclamation_T']
X1_train = df_train[predictors]
X1_test = df_test[predictors]
y_train = df_train['Score']
y_test = df_test['Score']

In [ ]:
from sklearn import neighbors     

# choose model class
n_neighbors = 5
# instantiate model
model = neighbors.KNeighborsRegressor(n_neighbors) 
# fit model to data    
model.fit(X1_train, y_train)                            

KNeighborsRegressor()

In [ ]:
# predict on training data
print("predictions on training set:")
y_train_fit = model.predict(X1_train)        
mse_train = np.mean( (y_train - y_train_fit)**2 )
print(np.sqrt(mse_train), mse_train)    

predictions on training set:
0.8285365278829268 0.6864727780362959


In [ ]:
# predict on test data
print("predictions on test set:")
y_test_fit = model.predict(X1_test)                     
mse_test = np.mean( (y_test - y_test_fit)**2 )
print(np.sqrt(mse_test), mse_test)

predictions on test set:
1.0312208203376343 1.0634163802978236


This means that, on average, the model's predictions are off by 0.6864727780362959 for the training set and 1.0634163802978236 for the testing set. This indicates that the model performs better on the training set than on the testing set. But, since the difference between the RMSEs is not very large, which suggests that the model is not overfitting the training data.